### PR TITLE
Add browserkit component documentation

### LINKED ISSUE #954: Bootstrap BrowserKit component documentation
Recommended sections:

* Introduction
* Installation
* Usage

### PR DESCRIPTION
**Pull request by yamiko-ninja**
Working on documentation for the browser kit. Fixes #954
Pull request #4310 was to the wrong branch so making a new one

### CODE REVIEW

**Comment by WouterJ** [comment]
ping @yamiko-ninja do you have time in the future to bring this PR to the next level? If not, please say so, then someone else can take it over.

**Comment by yamiko-ninja** [comment]
@WouterJ ahh forgot about it! 

If someone wants to take over they can. otherwise I will put this on my todo list to do by the end of April. About to move in a few weeks.

**Comment by WouterJ** [comment]
Hi @yamiko-ninja! Just wanted to let you know that there will be a [doc sprint day](http://symfony.com/blog/symfony-docs-hack-day-needs-you-on-may-23rd) on May 23rd. You can join in then to get quick feedback & other help. If you no longer have the motivation to finish this PR, we can also put it on the list for other people to take over.

Thanks for the bootstrapping!

**Comment by yamiko-ninja** [comment]
I will put that day on my calendar and will do some more work on it in the meantime. I have time to work on this again now.

**Comment by javiereguiluz** [comment]
@WouterJ I have some concerns about this PR. @yamiko-ninja has done a great work bootstraping the documentation of the component. However, if we expect him to complete all the documentation, maybe that's too much for a single person (hence my concerns).

What do you think if we lower the expectations of this PR to consider it just the bootstraping of the component documentation. That way we can polish @yamiko-ninja's current work a bit and merge it. Then, we outline all the contents we want to add to this section and complete that task in further pull requests.

**Comment by xabbuh** [comment]
I agree with you @javiereguiluz for two reasons:

1. Only bootstrapping a rough skeleton would make it easier to get something started.
1. We can then create a list of smaller TODOs which would make it easier for other contributors to pick a single simple task.

**Comment by yamiko-ninja** [comment]
I agree. I have only made a few contributions here and there I would be fine with someone else taking over.

File: components/browser_kit/introduction.rst [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+Creating a Client
+-----------------
+
+To create your own client you must extend the abstract client class and implement the doRequest method.
+This method accepts a request and should return a response.
+
+.. code-block:: php
+
+    namespace ACME;
```
Comment (OskarStark) [inline components/browser_kit/introduction.rst]: should be ``AppBundle``

Comment (WouterJ) [inline components/browser_kit/introduction.rst]: No, we are in the components section here. It should not use framework things like bundles

**Comment by yamiko-ninja** [comment]
I added in some docs on cookies. I think this is a good start and other can add more to it if they like. With that in mind I removed the [WIP] from the title.

File: components/browser_kit/introduction.rst [hunk 2]
```diff
@@ -0,0 +1,233 @@
+.. index::
+   single: BrowserKit
+   single: Components; BrowserKit
+
+The BrowserKit Component
+========================
+
+    The BrowserKit component simulates the behavior of a web browser.
+
+The BrowserKit component allows you to make web request, click on links and submit forms.
```
Comment (xabbuh) [inline components/browser_kit/introduction.rst]: requests

File: components/browser_kit/introduction.rst [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+========================
+
+    The BrowserKit component simulates the behavior of a web browser.
+
+The BrowserKit component allows you to make web request, click on links and submit forms. 
+
+Installation
+------------
+
+You can install the component in 2 different ways:
```
Comment (xabbuh) [inline components/browser_kit/introduction.rst]: "two" instead of "2"

File: components/browser_kit/introduction.rst [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+* :doc:`Install it via Composer </components/using_components>` (``symfony/browser-kit`` on `Packagist`_);
+* Use the official Git repository (https://github.com/symfony/BrowserKit).
+
+Basic Usage
+-----------
+
+.. note::
+    The component only provides an abstract client and does not provide any "default" backend for the HTTP layer.
```
Comment (xabbuh) [inline components/browser_kit/introduction.rst]: There should be a blank line between the directive and its contents.

File: components/browser_kit/introduction.rst [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+This method accepts a request and should return a response.
+
+.. code-block:: php
+
+    namespace ACME;
+
+    use Symfony\Component\BrowserKit\Client as BaseClient;
+    use Symfony\Component\BrowserKit\Response;
+
+    class Client extends BaseClient {
```
Comment (xabbuh) [inline components/browser_kit/introduction.rst]: We always put the brace on the next line (same for methods).

File: components/browser_kit/introduction.rst [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        protected function doRequest($request) {
+            // convert request into a response
+            // ...
+            return new Response($content, $status, $headers);
+        }
+    }
+
+For a simple implementation of a browser based on an HTTP layer, have a look at Goutte_.
+
+For an implementation based on HttpKernelInterface, have a look at the Client provided by the :doc:`/components/http_kernel/introduction`.
```
Comment (xabbuh) [inline components/browser_kit/introduction.rst]: You should enclose the interface name with literals (double backticks).

File: components/browser_kit/introduction.rst [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+
+Making Request
```
Comment (xabbuh) [inline components/browser_kit/introduction.rst]: Requests

File: components/browser_kit/introduction.rst [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+~~~~~~~~~~~~~~
+
+To make a request you use the client's request method.
```
Comment (xabbuh) [inline components/browser_kit/introduction.rst]: You can use the `method` role to link to the API docs of the `request()` method here.

File: components/browser_kit/introduction.rst [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+For a simple implementation of a browser based on an HTTP layer, have a look at Goutte_.
+
+For an implementation based on HttpKernelInterface, have a look at the Client provided by the :doc:`/components/http_kernel/introduction`.
+
+
+Making Request
+~~~~~~~~~~~~~~
+
+To make a request you use the client's request method. 
+The first two arguments are for the HTTP method and the request url.
```
Comment (xabbuh) [inline components/browser_kit/introduction.rst]: URL

File: components/browser_kit/introduction.rst [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+    use ACME\Client;
+
+    $client = new Client();
+    $crawler = $client->request('GET', 'http://symfony.com');
+
+Clicking Links
+~~~~~~~~~~~~~~
+
+Select a link with the crawler and pass it to the click method to click on the link.
```
Comment (xabbuh) [inline components/browser_kit/introduction.rst]: Same here, you can also link to the API docs of the `click()` method.

File: components/browser_kit/introduction.rst [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+    );
+
+    // create a client and set the cookies
+    $client = new Client(array(), array(), $cookieJar);
+    // ...
+
+History
+-------
+
+The client stores all your request allowing you to go back and forward in your history.
```
Comment (xabbuh) [inline components/browser_kit/introduction.rst]: requests

File: components/browser_kit/introduction.rst [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+    $doc_link = $crawler->selectLink('Documentation')->link();
+    $doc_crawler = $client->click($link);
+
+    // go back to home page
+    $home_crawler = $client->back();
+
+    // go forward to documentation page
+    $doc_crawler = $client->forward();
+
+You can restart the clients history with the restart method. This will also clear out the CookieJar.
```
Comment (xabbuh) [inline components/browser_kit/introduction.rst]: [...] the client's history [...]

**Comment by xabbuh** [comment]
This looks pretty good so far. Thanks for your work @yamiko-ninja! I have left you a few comments (I wasn't able to look into this too deeply yet). And could you please also update this so that each lines is wrapped after the first word that crosses the 72nd character?

**Comment by yamiko-ninja** [comment]
@xabbuh thanks for all the corrections. I will make the changes over the weekend.

**Comment by javiereguiluz** [comment]
@yamiko-ninja you did a great work in this pull request. We just need a final push to finish it. That's what I'm trying to do in #6072. I've reused your original commits, so you'll get full credit for your work once the other pull request is merged. Thanks!

**Pull request closed without merging**